### DISCUSSION TITLE
Can't read env variable set after build

### DISCUSSION THREAD

**Discussion by brunouber** [vercel/next.js#52543] [App Router] [answered]
Is there a way to make Next.js read env variables set after the build is complete?

```
    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: 1.17.3
      pnpm: N/A
    Relevant Packages:
      next: 13.4.9
      eslint-config-next: 13.4.9
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
    Next.js Config:
      output: N/A
```

In Next.js route, can't read environment variables set after build.

I've created a simple route, like this:

```
// app/config/route.ts

import { NextResponse } from 'next/server';

export const GET = () => {
    return NextResponse.json({
        value: process.env.MY_VALUE,
    });
};
```

Then I run the commands:

```
npm run build
export MY_VALUE="Hello world!"
echo $MY_VALUE
npm start
```

When i GET /config `value` is not set.

**Comment by icyJoseph** [comment, marked as answer]
Hi, 

When your build your app, there's a summary, could you tell me what the `app/config/routes.ts` summary? I am guessing it is statically optimised.

- https://nextjs.org/docs/app/building-your-application/routing/router-handlers#static-route-handlers

Try putting a `export const dynamic = 'force-dynamic'` above the `export const GET` declaration

**Reply by brunouber** [reply]
Adding `export const dynamic = 'force-dynamic'` fixes the issue.
Another thing I tested before reading your answer is moving the api endpoint in the old `pages` folder, and that fixes the issue as well - but i think your solution is preferable.

Thank you

**Reply by icyJoseph** [reply]
yeah, the thing is that, whenever possible, Route handlers are getting optimised, or rather statically rendered, and you have to opt them into being dynamic. A bit unintuitive behaviour, I think.

**Reply by brunouber** [reply]
Do you know whether there's a way to get dynamic value also of env variables that should be available to frontend (eg. NEXT_PUBLIC_SOMETHING)?

**Reply by icyJoseph** [reply]
No, those as far as I remember, are inlined into the JavaScript bundle.
